src/new/wallet.js

```diff
@@ -2666,6 +2675,168 @@ class HathorWallet extends EventEmitter {
+    }
+
+    // Here we validate that the arguments sent in the data array of args has
+    // the expected type for each parameter of the blueprint method
+    for (const [index, arg] of methodArgs.entries()) {
+      let typeToCheck = arg.type;
+      if (typeToCheck.startsWith('SignedData')) {
+        // Signed data will always be an hexadecimal with the
+        // signed part and the data itself
+        typeToCheck = 'str';
+      }
+      switch (typeToCheck) {
+        case 'bytes':
+          // Bytes arguments are sent in hexadecimal
+          try {
+            data.args[index] = hexToBuffer(data.args[index]);
+          } catch {
+            // Data sent is not a hex
+            throw new NanoContractTransactionError(`Invalid hexadecimal for argument number ${index + 1}.`);
+          }
+          break;
+        case 'int':
+        case 'float':
+          if (typeof data.args[index] !== 'number') {
+            throw new NanoContractTransactionError(`Expects argument number ${index + 1} type ${arg.type} but received type ${typeof data.args[index]}.`);
+          }
+          break;
+        case 'str':
+          if (typeof data.args[index] !== 'string') {
+            throw new NanoContractTransactionError(`Expects argument number ${index + 1} type ${arg.type} but received type ${typeof data.args[index]}.`);
+          }
+          break;
+        default:
+          if (arg.type !== typeof data.args[index]) {
+            throw new NanoContractTransactionError(`Expects argument number ${index + 1} type ${arg.type} but received type ${typeof data.args[index]}.`);
+          }
+      }
+    }
+
+    // Get private key that will sign the nano contract transaction
+    let derivedKey;
+    try {
+      derivedKey = await this.getHDPrivateKeyFromAddress(address, options);
+    } catch (e) {
+      if (e instanceof AddressError) {
+        throw new NanoContractTransactionError('Address used to sign the transaction does not belong to the wallet.');
+      }
+      throw e;
+    }
+    const privateKey = derivedKey.privateKey;
+
+    // Build and send transaction
+    const builder = new NanoContractTransactionBuilder();
+    builder.setMethod(method);
+    builder.setWallet(this);
+    builder.setBlueprintId(blueprintId);
+    builder.setNcId(method === 'initialize' ? null : data.ncId);
+    builder.setCaller(privateKey);
+    builder.setActions(data.actions);
+    builder.setArgs(data.args);
+
+    const nc = await builder.build();
+    return signAndPushNCTransaction(nc, privateKey, pin, this.storage);
+  }
+
+  /**
+   * Generate and return the HDPrivateKey for an address
+   *
+   * @param {string} address Address to get the HDPrivateKey from
+   * @param [options]
+   * @param {string} [options.pinCode] PIN to decrypt the private key.
+   *                                   Optional but required if not set in this
+   *
+   * @returns {Promise<HDPrivateKey>}
+   */
+  async getHDPrivateKeyFromAddress(address, options = {}) {
```

Comment:
As stated on the other comment. maybe this should return the privateKey since address level does not require any derivation.
And from what I read we get the hdprivateKey and just use the privateKey anyway

Reply:
Makes sense, did the refactor. https://github.com/HathorNetwork/hathor-wallet-lib/pull/604/commits/e57485b815213e3e9f17a94f641f61a59c532277